src/Microsoft.AspNetCore.Session/DistributedSession.cs

```diff
@@ -128,43 +166,64 @@ public void Set(string key, byte[] value)
             }
 
             Load();
-            if (!_tryEstablishSession())
+            if (IsAvailable)
             {
-                throw new InvalidOperationException(Resources.Exception_InvalidSessionEstablishment);
+                if (!_tryEstablishSession())
+                {
+                    throw new InvalidOperationException(Resources.Exception_InvalidSessionEstablishment);
+                }
+                _isModified = true;
+                byte[] copy = new byte[value.Length];
+                Buffer.BlockCopy(src: value, srcOffset: 0, dst: copy, dstOffset: 0, count: value.Length);
+                _store[encodedKey] = copy;
             }
-            _isModified = true;
-            byte[] copy = new byte[value.Length];
-            Buffer.BlockCopy(src: value, srcOffset: 0, dst: copy, dstOffset: 0, count: value.Length);
-            _store[encodedKey] = copy;
         }
 
         public void Remove(string key)
         {
             Load();
-            _isModified |= _store.Remove(new EncodedKey(key));
+            if (IsAvailable)
+            {
+                _isModified |= _store.Remove(new EncodedKey(key));
+            }
         }
 
         public void Clear()
         {
             Load();
-            _isModified |= _store.Count > 0;
-            _store.Clear();
+            if (IsAvailable)
+            {
+                _isModified |= _store.Count > 0;
+                _store.Clear();
+            }
         }
 
         private void Load()
         {
             if (!_loaded)
             {
-                var data = _cache.Get(_sessionKey);
-                if (data != null)
+                try
                 {
-                    Deserialize(new MemoryStream(data));
+                    var data = _cache.Get(_sessionKey);
+                    if (data != null)
+                    {
+                        Deserialize(new MemoryStream(data));
+                    }
+                    else if (!_isNewSessionKey)
+                    {
+                        _logger.AccessingExpiredSession(_sessionKey);
+                    }
+                    _isAvailable = true;
                 }
-                else if (!_isNewSessionKey)
+                catch (Exception exception)
                 {
-                    _logger.AccessingExpiredSession(_sessionKey);
+                    _logger.SessionCacheReadException(_sessionKey, exception);
+                    _isAvailable = false;
+                }
+                finally
+                {
+                    _loaded = true;
                 }
-                _loaded = true;
             }
         }
 
```

Comment:
I think we can remove `TODO` item on the like below since that's what we will go with. Same with the method `Deserialize` where we will not throw since it's self healing by creating a new session. `ReadBytes` may still fail due to corrupted cache but that exception will be caught in `Load`.